notation.go

```diff
@@ -63,31 +44,189 @@ type SignOptions struct {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if err != nil {
+		if outcome == nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+	} else if outcome.VerificationLevel.Name == trustpolicy.LevelSkip.Name {
+		return ocispec.Descriptor{}, []*VerificationOutcome{outcome}, nil
+	}
+
+	// get signature manifests
+	var targetArtifactDesc ocispec.Descriptor
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+	count := 0
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		if count >= verificationSignatureNumLimit {
+			return fmt.Errorf("number of signatures attempted to be verified should be less than: %d", verificationSignatureNumLimit)
+		}
+		// process signatures
+		for _, sigManifestDesc := range signatureManifests {
+			// get signature envelope
+			sigBlob, _, err := repo.FetchSignatureBlob(ctx, sigManifestDesc)
+			if err != nil {
+				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %v", sigManifestDesc.Digest, artifactRef, err.Error())}
+			}
+			payloadArtifactDescriptor, outcome, err := verifier.Verify(ctx, sigBlob, opts)
+			if err != nil {
+				if outcome == nil {
+					// TODO: log fatal error
+					return err
+				}
+				verificationOutcomes = append(verificationOutcomes, outcome)
+				continue
+			}
+			if !content.Equal(payloadArtifactDescriptor, artifactDescriptor) {
+				outcome.Error = errors.New("content descriptor mismatch")
+				verificationOutcomes = append(verificationOutcomes, outcome)
+				continue
+			}
+
+			// At this point, we've found a signature verified successfully
+			outcome.RawSignature = sigBlob
+			verificationOutcomes = append(verificationOutcomes, outcome)
+			// Descriptor of the signature blob that gets verified successfully
+			targetArtifactDesc = payloadArtifactDescriptor
+
+			return errDoneVerification
+		}
+		return nil
+	})
+
+	if err != nil && !errors.Is(err, errDoneVerification) {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	// If there's no signature associated with the reference
+	if len(verificationOutcomes) == 0 {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("no signature is associated with %q, make sure the image was signed successfully", artifactRef)}
+	}
+
+	// check whether verification was successful or not
+	if verificationOutcomes[len(verificationOutcomes)-1].Error == nil {
+		// signature verification succeeds if there is at least one good
+		// signature
+		return targetArtifactDesc, verificationOutcomes, nil
+	}
+
+	return ocispec.Descriptor{}, verificationOutcomes, ErrorVerificationFailed{}
```

Comment:
nit: code style check

Instead of doing
```go
err := do()
if err == nil {
    return process()
}
return err
```
try
```go
err := do()
if err != nil {
    return err
}
process()
```

```suggestion
	// check whether verification was successful or not
	if verificationOutcomes[len(verificationOutcomes)-1].Error != nil {
		return ocispec.Descriptor{}, verificationOutcomes, ErrorVerificationFailed{}
	}
	return targetArtifactDesc, verificationOutcomes, nil
```